### PR TITLE
Fix AsyncResolver to match ThreadedResolver behavior

### PR DESCRIPTION
**Pull request by bdraco**
## What do these changes do?

AsyncResolver was disabled by default because it did not implement all of functionality of ThreadedResolver because `aiodns` did not support `getaddrinfo` until https://github.com/saghul/aiodns/pull/118 which meant only IPv6 addresses were returned in most cases on systems that supported IPv6

see https://github.com/aio-libs/aiohttp/issues/559

## Are there changes in behavior for the user?

`AsyncResolver` now behaves the same as the current default `ThreadedResolver` and will return both IPv6 and IPv4 addresses on systems that support it.

## Is it a substantial burden for the maintainers to support this?

no

## Checklist

- [ ] I think the code is well written
- [ ] Unit tests for the changes exist
- [ ] Documentation reflects the changes
- [ ] If you provide code modification, please add yourself to `CONTRIBUTORS.txt`
  * The format is &lt;Name&gt; &lt;Surname&gt;.
  * Please keep alphabetical order, the file is sorted by names.
- [ ] Add a new news fragment into the `CHANGES/` folder
  * name it `<issue_or_pr_num>.<type>.rst` (e.g. `588.bugfix.rst`)
  * if you don't have an issue number, change it to the pull request
    number after creating the PR
    * `.bugfix`: A bug fix for something the maintainers deemed an
      improper undesired behavior that got corrected to match
      pre-agreed expectations.
    * `.feature`: A new behavior, public APIs. That sort of stuff.
    * `.deprecation`: A declaration of future API removals and breaking
      changes in behavior.
    * `.breaking`: When something public is removed in a breaking way.
      Could be deprecated in an earlier release.
    * `.doc`: Notable updates to the documentation structure or build
      process.
    * `.packaging`: Notes for downstreams about unobvious side effects
      and tooling. Changes in the test invocation considerations and
      runtime assumptions.
    * `.contrib`: Stuff that affects the contributor experience. e.g.
      Running tests, building the docs, setting up the development
      environment.
    * `.misc`: Changes that are hard to assign to any of the above
      categories.
  * Make sure to use full sentences with correct case and punctuation,
    for example:
    ```rst
    Fixed issue with non-ascii contents in doctest text files
    -- by :user:`contributor-gh-handle`.
    ```

    Use the past tense or the present tense a non-imperative mood,
    referring to what's changed compared to the last released version
    of this project.

needs 
- [x] https://github.com/saghul/aiodns/pull/118
- [x] https://github.com/saghul/aiodns/pull/119
- [x] aiodns release (requested at https://github.com/saghul/aiodns/pull/119#issuecomment-2028487751)

### CODE REVIEW

File: aiohttp/resolver.py [hunk 1]
```diff
@@ -85,19 +85,49 @@ def __init__(self, *args: Any, **kwargs: Any) -> None:
...
+            if family == socket.AF_INET6:
+                if len(address) < 3:
+                    # IPv6 is not supported by Python build,
+                    # or IPv6 is not enabled in the host
+                    continue
+                if address[3]:
+                    # This is essential for link-local IPv6 addresses.
+                    # LL IPv6 is a VERY rare case. Strictly speaking, we should use
+                    # getnameinfo() unconditionally, but performance makes sense.
+                    host, _port = await self._loop.getnameinfo(
```
Comment (bdraco) [inline aiohttp/resolver.py]: pycares supports getnameinfo so we can avoid the executor here as well but I need to do another PR to aiodns

File: aiohttp/resolver.py [hunk 2]
```diff
@@ -85,19 +85,49 @@ def __init__(self, *args: Any, **kwargs: Any) -> None:
         self._resolver = aiodns.DNSResolver(*args, loop=self._loop, **kwargs)
 
     async def resolve(
-        self, host: str, port: int = 0, family: int = socket.AF_INET
+        self, hostname: str, port: int = 0, family: int = socket.AF_INET
```
Comment (Dreamsorcerer) [inline aiohttp/resolver.py]: The parent class names this `host`, so should probably keep that name (not sure why mypy doesn't give an error about this...).

Comment (bdraco) [reply aiohttp/resolver.py]: I was matching it to ThreadedResolver. I guess we change both to host?

Comment (Dreamsorcerer) [reply aiohttp/resolver.py]: Probably makes more sense.

**Comment by bdraco** [comment]
I've tested this on production and everything works as expected.  I'll do another turn and address review comments after the aiodns release

**Comment by bdraco** [comment]
Functional testing looks good, but all the tests for the resolver are patching the old Cath so lots to update

File: aiohttp/resolver.py [hunk 3]
```diff
@@ -48,19 +48,19 @@ async def resolve(
                     # This is essential for link-local IPv6 addresses.
                     # LL IPv6 is a VERY rare case. Strictly speaking, we should use
                     # getnameinfo() unconditionally, but performance makes sense.
-                    host, _port = socket.getnameinfo(
+                    resolved_host, _port = socket.getnameinfo(
```
Comment (bdraco) [inline aiohttp/resolver.py]: Looks like there is no coverage for this ?

Comment (bdraco) [reply aiohttp/resolver.py]: Also its using socket and not loop!

File: tests/test_resolver.py [hunk 4]
```diff
@@ -49,21 +93,77 @@ async def fake(*args: Any, **kwargs: Any) -> List[Any]:
...
+        mock().getaddrinfo.assert_called_with(
+            "www.python.org",
+            family=socket.AF_INET,
+            flags=socket.AI_ADDRCONFIG,
+            port=0,
+            type=socket.SOCK_STREAM,
+        )
+
+
 @pytest.mark.skipif(not gethostbyname, reason="aiodns 1.1 required")
```
Comment (bdraco) [inline tests/test_resolver.py]: This needs to be getaddrinfo as well as the other ones, but has to wait for a real release to change

File: aiohttp/resolver.py [hunk 5]
```diff
@@ -48,24 +51,24 @@ async def resolve(
                     # This is essential for link-local IPv6 addresses.
                     # LL IPv6 is a VERY rare case. Strictly speaking, we should use
                     # getnameinfo() unconditionally, but performance makes sense.
-                    host, _port = socket.getnameinfo(
-                        address, socket.NI_NUMERICHOST | socket.NI_NUMERICSERV
+                    resolved_host, _port = await self._loop.getnameinfo(
```
Comment (bdraco) [inline aiohttp/resolver.py:56]: This has apparently never worked on python less than 3.9

Comment (bdraco) [reply aiohttp/resolver.py:56]: Yeah, it's the same problem we have in zeroconf where scope ID isn't supported on an IP address until 3.9.

Sisters never been in any test coverage here, the problem has gone unnoticed since it's very rare. This is even used anyway nobody seems to have complained

Comment (Dreamsorcerer) [reply aiohttp/resolver.py:56]: Huh, oh well, we'll drop support for 3.8 in a few months anyway.

Comment (bdraco) [reply aiohttp/resolver.py:56]: I'm going to mark the test as skip on 3.8 as I don't think there's any point in trying to fix it as it simply isn't going to work

I doubt anyone's using it on 3.8 who would even need this... I guess they really can't anyways

File: aiohttp/abc.py [hunk 6]
```diff
@@ -117,11 +119,23 @@ def __await__(self) -> Generator[Any, None, StreamResponse]:
         """Execute the view handler."""
 
 
+class ResolveResult(TypedDict):
```
Comment (webknjaz) [inline aiohttp/abc.py:122]: add a docstring?

**Comment by webknjaz** [comment]
@bdraco does this need to include a minimal version bump in `setup.cfg` or is it API-compatible?

**Comment by bdraco** [comment]
It's going to need aiodns bumped.  It was just released so I'll get that done after breakfast

**Comment by bdraco** [comment]
I would like to backport this fix to 3.9 as well since AsyncResolver is currently unusable without it (currently if the system supports IPv6 and there are no AAAA records AsyncResolver doesn't fallback to A in many cases)

Ideally we can switch the default back to AsyncResolver now that the semantics should be the same as ThreadedResolver but I think that change should wait until 3.10

**Comment by Dreamsorcerer** [comment]
> I would like to backport this fix to 3.9

Are we going to break anything if someone is using an older aiodns version? I'd prefer to avoid changing the dependency requirements in a patch version.

**Comment by bdraco** [comment]
> > I would like to backport this fix to 3.9
> 
> 
> 
> Are we going to break anything if someone is using an older aiodns version? I'd prefer to avoid changing the dependency requirements in a patch version.

Yes it would. aiodns does not expose the needed API until the release done earlier today.

File: aiohttp/abc.py (hunk 6 above)
Comment (bdraco) [reply aiohttp/abc.py:122]: Good call. added

File: CHANGES/8270.bugfix.rst [hunk 7]
```diff
@@ -0,0 +1,9 @@
+Fix AsyncResolver to match ThreadedResolver behavior
+-- by :user:`bdraco`.
+
+On system with IPv6 support, the AsyncResolver would not fallback
+to providing A records when AAAA records were not available.
+Additionally, unlikely the ThreadedResolver, the AsyncResolver
```
Comment (bdraco) [inline CHANGES/8270.bugfix.rst]:
```suggestion
Additionally, unlike the ThreadedResolver, the AsyncResolver
```

**Comment by webknjaz** [comment]
> > I would like to backport this fix to 3.9
> 
> Are we going to break anything if someone is using an older aiodns version? I'd prefer to avoid changing the dependency requirements in a patch version.

Technically, if *we* don't change our API, it's not a reason for us to include this in a minor or major version bump. SemVer-wise, that is. Fulfilling the dependencies is a job of a dependency resolver/installer. Similar to setting a specific `python_requires` value that wouldn't be an API change and wouldn't be breaking from this perspective.

File: CHANGES/8270.bugfix.rst [hunk 8]
```diff
@@ -0,0 +1,9 @@
+Fix AsyncResolver to match ThreadedResolver behavior
+-- by :user:`bdraco`.
+
+On system with IPv6 support, the AsyncResolver would not fallback
+to providing A records when AAAA records were not available.
+Additionally, unlike the ThreadedResolver, the AsyncResolver
```
Comment (webknjaz) [inline CHANGES/8270.bugfix.rst]: Could you double-backtick the identifiers?

File: aiohttp/abc.py (hunk 6 above)
Comment (webknjaz) [reply aiohttp/abc.py:122]: Perhaps, we should also link it from `abc.rst`?

File: CHANGES/8270.bugfix.rst (hunk 8 above)
Comment (bdraco) [reply CHANGES/8270.bugfix.rst]: 4f492651b3ae3e693cbbf33916fe05d73e2d046e

File: CHANGES/8270.bugfix.rst [hunk 9]
```diff
@@ -0,0 +1,9 @@
+Fix ``AsyncResolver`` to match ``ThreadedResolver`` behavior
+-- by :user:`bdraco`.
+
+On system with IPv6 support, the :py:class:`~aiohttp.resolver.AsyncResolver` would not fallback
```
Comment (webknjaz) [inline CHANGES/8270.bugfix.rst:4]: Is it linkable? Looking at https://webknjaz.github.io/intersphinx-untangled/docs.aiohttp.org/, it doesn't appear so. You'd need to add it to https://github.com/aio-libs/aiohttp/blob/1542348/docs/conf.py#L396 or document in RST. Otherwise, I expect Sphinx to choke: https://github.com/aio-libs/aiohttp/actions/runs/8499653747/job/23280716056?pr=8270#step:13:129.

Comment (bdraco) [reply CHANGES/8270.bugfix.rst:4]: Thanks. Sadly it looks like they are undocumented so 3d5175f2ddeec003d01793a4c7bd7319c7abc25d

Comment (webknjaz) [reply CHANGES/8270.bugfix.rst:4]: Yeah, that's why I initially suggested double backticks as this is the easiest way to highlight the identifiers without having to deal with Sphinx. Marking as resolved.

File: docs/conf.py [hunk 10]
```diff
@@ -394,6 +394,8 @@
     ("py:class", "aiohttp.ClientRequest"),  # undocumented
     ("py:class", "aiohttp.payload.Payload"),  # undocumented
     ("py:class", "aiohttp.abc.AbstractResolver"),  # undocumented
+    ("py:class", "aiohttp.abc.AsyncResolver"),  # undocumented
+    ("py:class", "aiohttp.abc.ThreadedResolver"),  # undocumented
```
Comment (webknjaz) [inline docs/conf.py]:
```suggestion
    ("py:class", "aiohttp.resolver.AsyncResolver"),  # undocumented
    ("py:class", "aiohttp.resolver.ThreadedResolver"),  # undocumented
```

Comment (bdraco) [reply docs/conf.py]: Thanks 👍

File: aiohttp/abc.py (hunk 6 above)
Comment (bdraco) [reply aiohttp/abc.py:122]: It looks like the `AbstractResolver` isn't mentioned at all in `abc.rst`

Comment (bdraco) [reply aiohttp/abc.py:122]: I added it to `abc.rst`

**Comment by Dreamsorcerer** [comment]
> > > I would like to backport this fix to 3.9
> > 
> > 
> > Are we going to break anything if someone is using an older aiodns version? I'd prefer to avoid changing the dependency requirements in a patch version.
> 
> Technically, if _we_ don't change our API, it's not a reason for us to include this in a minor or major version bump. SemVer-wise, that is. Fulfilling the dependencies is a job of a dependency resolver/installer. Similar to setting a specific `python_requires` value that wouldn't be an API change and wouldn't be breaking from this perspective.

Sure, I just think it's preferable to minimise requirement changes in order to get security/bug fixes. I'll let you two decide, I don't mind too much.

**Comment by bdraco** [comment]
The interface is flexible enough that anyone who needs to fix can copy the code until it's released so I'm fine with leaving it in the current state for 3.9. 

It looks like anyone using it already has a workaround in place of setting AF_INET instead of unspecified so they have been living with this issue for years anyways. That's not great for IPv6 adoption but they also have ThreadedResolver even if its performance is worse so there are viable workarounds here now.

File: docs/conf.py [hunk 11]
```diff
@@ -394,6 +394,8 @@
     ("py:class", "aiohttp.ClientRequest"),  # undocumented
     ("py:class", "aiohttp.payload.Payload"),  # undocumented
     ("py:class", "aiohttp.abc.AbstractResolver"),  # undocumented
```
Comment (webknjaz) [inline docs/conf.py]: Since you're documenting it, the ignore shouldn't be needed anymore:
```suggestion
```

**Comment by bdraco** [comment]
Got distracted yesterday with the HA release so I didn't have time to get back to this.

I'm going to push this to another production system to be sure everything is OK shortly before merging

**Comment by bdraco** [comment]
So far so good on both production systems.

Will check it again after dinner before merging.

**Comment by bdraco** [comment]
All good

Also good with IPv6 disabled in the kernel.

I can't think of any other way to beat this up. So I'm happy with it now
